### PR TITLE
Add Member During Network Partition

### LINKED ISSUE #54: Do we need to remove node on failure?
From https://github.com/edegens/ptype/issues/15#issuecomment-591069320 we saw that changes to the etcd cluster won't be allowed if it isn't healthy. The case that we are concerned about is if a node in the cluster gets taken down. We think it might jeopardize the health of the cluster and not allow changes. 

If we do need to remove it on failure, we need to build a mechanism for it.

### PR DESCRIPTION
**Pull request by JackyChiu**
closes #54 

# What

This PR covers the research and features needed to add members during network partitions and how to do it safely. This is further proved in the tests. 

# How

The root problem comes from the fact that when `strict-reconfig-check` is enabled, all etcd nodes must be [connected](https://github.com/etcd-io/etcd/issues/10738) even though the cluster [should only require quorum](https://github.com/etcd-io/etcd/issues/10114) to do so. The reason is to avoid the edge case where a newly introduce node is misconfigured, where it can result in a loss of quorum since the new node counts as a voting node right away. 

This PR leaves the `strict-reconfig-check` up to the developers to disable themselves. However I did introduce a feature to reduce the risk when adding a new node. 

In v3.4 etcd introduced a [learner](https://etcd.io/docs/v3.3.12/learning/learner/#features-in-v34) feature. This is an intermediate state used where a node is added as a member but doesn't count as a voter. Once it caught up in the leader's logs, it can be promoted to a voter node. This is where the cluster checks if the node is healthy and configured to join, something that isn't done currently when added a new node. 

In this PR we add new members as learners and auto-promote them when they are caught up. The auto promote had to be built by us since it's [not available yet](https://github.com/etcd-io/etcd/pull/10887).

# Review

Review by commit for an easier time. 

# tldr
- disable `strict-reconfig-check` to add node during network partition
- adding members as learners reduces the risk of losing quorum

### CODE REVIEW

File: cluster/cluster_test.go [hunk 1]
```diff
@@ -87,44 +79,31 @@ func (suite *ClusterSuite) TestMemberAdd() {
...
+		memberCfg.etcdConfig.APUrls = []url.URL{*LPUrl}
+		memberCfg.etcdConfig.ACUrls = []url.URL{*LCUrl}
 		memberCfg.etcdConfig.ClusterState = embed.ClusterStateFlagExisting
 
 		c2, err := Join(ctx, memberCfg)
 		require.NoError(t, err)
 		defer c2.Close()
 
-		members, err := c.MemberList(ctx)
+		members, err := c2.MemberList(ctx)
```
Comment (bkzhang) [inline cluster/cluster_test.go]: This was initially to make sure that the initial cluster got another member, if Join ever changes then we still want to ensure that `c` gets this node added because `c2` here might end up being a different cluster if it bootstraps a new cluster or something if logic changes (and if this makes a new cluster of 1 and not 2, someone might just change the `MemberList` check below to equal 1). We can get both `c` and `c2` if you want but I think we should keep a check for `c`

File: cluster/cluster_test.go [hunk 2]
```diff
@@ -136,19 +115,54 @@ func (suite *ClusterSuite) TestMemberAdd() {
...
+		memberCfg.etcdConfig.APUrls = []url.URL{*LPUrl, *LPUrl2}
+		memberCfg.etcdConfig.ACUrls = []url.URL{*LCUrl, *LCUrl2}
 		memberCfg.etcdConfig.ClusterState = embed.ClusterStateFlagExisting
 
 		c3, err := Join(ctx, memberCfg)
 		require.NoError(t, err)
 		defer c3.Close()
 
-		members, err = c.MemberList(ctx)
+		members, err = c3.MemberList(ctx)
```
Comment (bkzhang) [inline cluster/cluster_test.go]: same for this

File: cluster/cluster.go [hunk 3]
```diff
@@ -148,14 +152,38 @@ func urlsToString(urls []url.URL) []string {
...
+	}
+
+	initialClusterClient, err := clientv3.New(clientv3.Config{
+		Endpoints:   cfg.InitialClusterClientUrls,
+		DialTimeout: 5 * time.Second,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create etcd client from config: %w", err)
+	}
+	for {
```
[the remark this answers is not in the document]
Comment (JackyChiu) [reply cluster/cluster.go]: Ah I should make it time out. I was initially thinking that the context would do it for us but the context passed in most likely won't be set with a timeout

File: cluster/cluster_test.go (hunk 1 above)
Comment (JackyChiu) [reply cluster/cluster_test.go]: Makes sense! I'll change it back to just `c`

File: cluster/cluster.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bkzhang) [reply cluster/cluster.go]: yeah it'd be nice to have a separate error too so the user can handle the case without having to watch it for like 1 hour straight

[the remark this answers is not in the document]
Comment (JackyChiu) [reply cluster/cluster.go]: I'm letting the promotion call timeout with the new context provided

[the remark this answers is not in the document]
Comment (bkzhang) [reply cluster/cluster.go]: the server stays alive as a non learner member in this case right? is the server useful at all in this state?

[the remark this answers is not in the document]
Comment (JackyChiu) [reply cluster/cluster.go]: A time'd out promotion call will return an error and that will be propagated to the return value of the `Join` call. So it's up to the developer to handle what to do in that situation. In our examples we've just been considering it a fatal error

File: cluster/cluster_test.go [hunk 4]
```diff
@@ -149,6 +129,40 @@ func (suite *ClusterSuite) TestMemberAdd() {
...
+				etcdConfig:               embed.NewConfig(),
+			}
+			memberCfg.etcdConfig.Name = "node4"
+			memberCfg.etcdConfig.Dir = "tmp4"
+			memberCfg.etcdConfig.Logger = "zap"
+			memberCfg.etcdConfig.LPUrls = []url.URL{*LPUrl}
+			memberCfg.etcdConfig.LCUrls = []url.URL{*LCUrl}
+			memberCfg.etcdConfig.APUrls = []url.URL{*LPUrl}
+			memberCfg.etcdConfig.ACUrls = []url.URL{*LCUrl}
+			memberCfg.etcdConfig.ClusterState = embed.ClusterStateFlagExisting
```
Comment (dipshit) [inline cluster/cluster_test.go]: This is a great test! I think it is worth it to add a test with:

```golang
cfg.etcdConfig.StrictReconfigCheck = true
```

or

```yaml
strict-reconfig-check: false
```

for the cluster setup.

This is a decent amount of work, so if this won't regress then skip it I guess (as code changes are winding down).

Comment (JackyChiu) [reply cluster/cluster_test.go]: Thanks! In the tests, we've been using the `ping.yml` which ahs the `strict-reconfig-check: false` set. What test did you have in mind when setting it to true? That it won't allow for the added node?

Comment (dipshit) [reply cluster/cluster_test.go]: Yes. However after thinking about it, that is just testing an etcd feature, so nevermind.
